File: src/main/java/com/github/kelin/archetype/entity/Customer.java
package com.github.kelin.archetype.entity;

import static com.github.kelin.archetype.entity.EntityCollections.CUSTOMERS;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.ReadOnlyProperty;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.List;

@Document(collection = CUSTOMERS)
public class Customer {
    @Id
    public String id;

    public String firstName;
    public String lastName;
    public int position;

    @ReadOnlyProperty
    public List<CustomerRecord> records;

    @ReadOnlyProperty
    public CustomerExtra extra;

    public Customer() {
    }

    public Customer(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public Customer(String firstName, String lastName, int position) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.position = position;
    }

    public List<CustomerRecord> getRecords() {
        return records;
    }

    public void setRecords(List<CustomerRecord> records) {
        this.records = records;
    }

    public void setExtra(CustomerExtra extra) {
        this.extra = extra;
    }

    public CustomerExtra getExtra() {
        return extra;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
    }
}

File: src/main/java/com/github/kelin/archetype/entity/Customer.kt
package com.github.kelin.archetype.entity

import org.springframework.data.annotation.Id
import org.springframework.data.annotation.ReadOnlyProperty
import org.springframework.data.mongodb.core.mapping.Document

@Document(collection = EntityCollections.CUSTOMERS)
class Customer {
    @JvmField
    @Id
    var id: String? = null
    @JvmField
    var firstName: String? = null
    @JvmField
    var lastName: String? = null
    @JvmField
    var position = 0

    @JvmField
    @ReadOnlyProperty
    var records: List<CustomerRecord>? = null

    @JvmField
    @ReadOnlyProperty
    var extra: CustomerExtra? = null

    constructor()
    constructor(firstName: String?, lastName: String?) {
        this.firstName = firstName
        this.lastName = lastName
    }

    constructor(firstName: String?, lastName: String?, position: Int) {
        this.firstName = firstName
        this.lastName = lastName
        this.position = position
    }
}
